File: app/src/main/java/com/example/chatapp/MainActivity.java
package com.example.chatapp;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.net.Uri;
import android.os.Environment;
import android.os.PersistableBundle;
import android.os.Bundle;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.yslibrary.android.keyboardvisibilityevent.KeyboardVisibilityEvent;
import net.yslibrary.android.keyboardvisibilityevent.KeyboardVisibilityEventListener;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class MainActivity extends AppCompatActivity implements View.OnClickListener,View.OnFocusChangeListener,MessageAdapter.OnItemListener,EmojiAdapter.OnEmojiListener
{
    public static final int IMAGE_GALLERY_REQUEST = 20;
    private RecyclerView recyclerMsg,emojiPane;
    private MessageAdapter msgAdapter;
    private EmojiAdapter emojiAdapter;
    private RecyclerView.LayoutManager msgLayoutManager,emojiLayoutManager;
    com.wang.avi.AVLoadingIndicatorView firstType,secondType;
    EditText msg;
    static ImageButton sendBtn,attachmentBtn,menuBtn,item1,item2,item3,item4,item5;
    RelativeLayout menuBar,background,emojiRel;
    float d,SCREEN_WIDTH,SCREEN_HEIGHT;
    static int currentX,currentY,startX,startY,i=0,visible=0;
    boolean isEmojiPaneVisible=false;

    ArrayList<String> messages;
    ArrayList<Integer> sender;
    ArrayList<String> sentTime;
    ArrayList<String> sentDate;
    ArrayList<String> showdate;
    ArrayList<Integer> datePosition;
    static ArrayList<Integer> emojiName;
    RelativeLayout.LayoutParams menuBarParams;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        d = this.getResources().getDisplayMetrics().density;
        SCREEN_WIDTH=getWindowManager().getDefaultDisplay().getWidth();
        SCREEN_HEIGHT=getWindowManager().getDefaultDisplay().getHeight();

        messages = new ArrayList<String>();
        sender = new ArrayList<Integer>();
        sentTime = new ArrayList<String>();
        sentDate = new ArrayList<String>();
        showdate = new ArrayList<String>();
        datePosition = new ArrayList<Integer>();
        emojiName=new ArrayList<Integer>();



        msg=(EditText)findViewById(R.id.msg);
        sendBtn=(ImageButton)findViewById(R.id.sendBtn);
        recyclerMsg=(RecyclerView)findViewById(R.id.recyclerMsg);
        emojiPane=(RecyclerView)findViewById(R.id.emojiPane);
        firstType=(com.wang.avi.AVLoadingIndicatorView)findViewById(R.id.firstType);
        secondType=(com.wang.avi.AVLoadingIndicatorView)findViewById(R.id.secondType);
        attachmentBtn=(ImageButton)findViewById(R.id.attachmentBtn);
        menuBar=(RelativeLayout)findViewById(R.id.menuBar);
        menuBtn=(ImageButton)findViewById(R.id.menuBtn);
        background=(RelativeLayout)findViewById(R.id.background);
        item1=(ImageButton)findViewById(R.id.item1);
        item2=(ImageButton)findViewById(R.id.item2);
        item3=(ImageButton)findViewById(R.id.item3);
        item4=(ImageButton)findViewById(R.id.item4);
        item5=(ImageButton)findViewById(R.id.item5);
        emojiRel=(RelativeLayout)findViewById(R.id.emojiRel);

        initImages(this);
        emojiName=EmojiInitializer.initialize();

        firstType.setVisibility(View.GONE);
        secondType.setVisibility(View.GONE);
        menuBar.setVisibility(View.GONE);
        emojiRel.setVisibility(View.GONE);


        msgAdapter=new MessageAdapter(messages,sender,sentTime,showdate);
        msgAdapter.setListener(this);
        msgLayoutManager=new LinearLayoutManager(this);
        recyclerMsg.setLayoutManager(msgLayoutManager);
        recyclerMsg.setHasFixedSize(true);
        recyclerMsg.setAdapter(msgAdapter);
        msgLayoutManager.scrollToPosition(messages.size()-1);

        emojiAdapter=new EmojiAdapter(emojiName);
        emojiAdapter.setEmojiListener(this);
        emojiLayoutManager=new GridLayoutManager(this,5);
        emojiPane.setLayoutManager(emojiLayoutManager);
        emojiPane.setHasFixedSize(true);
        emojiPane.setAdapter(emojiAdapter);
        emojiLayoutManager.scrollToPosition(0);

        sendBtn.setOnClickListener(this);
        msg.setOnFocusChangeListener(this);
        attachmentBtn.setOnClickListener(this);
        menuBtn.setOnClickListener(this);
        background.setOnClickListener(this);
        recyclerMsg.setOnClickListener(this);
        item1.setOnClickListener(this);
        item2.setOnClickListener(this);
        item3.setOnClickListener(this);
        item4.setOnClickListener(this);
        item5.setOnClickListener(this);

        menuBarParams=(RelativeLayout.LayoutParams)menuBar.getLayoutParams();
        menuBarParams.addRule(RelativeLayout.ABOVE,R.id.baseBar);
        menuBarParams.setMargins(0,0,0,0);
        menuBar.setLayoutParams(menuBarParams);


        KeyboardVisibilityEvent.setEventListener(
                MainActivity.this,
                new KeyboardVisibilityEventListener() {
                    @Override
                    public void onVisibilityChanged(boolean isOpen)
                    {
                        if(isOpen==false)
                        {
                            View current = getCurrentFocus();
                            if (current != null) {current.clearFocus();}
                        }
                    }
                });


    }


    @Override
    public void onClick(View v)
    {
        if(v.equals(sendBtn))
        {
            if(visible==1) {
                visible = 0;
                menuBar.setVisibility(View.GONE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            }
            String message=msg.getText().toString().trim();
            if(message.length()>0)
            {
                messages.add(message);
                sender.add(i);
                sentTime.add(getTime());
                adjustDates();
                msgLayoutManager.scrollToPosition(messages.size()-1);
                msg.setText("");
                recyclerMsg.requestFocus();

                if(i==0)
                    i=1;
                else
                    i=0;

                msg.requestFocus();
            }
        }
        else if(v.equals(attachmentBtn))
        {
            if(visible==1) {
                visible = 0;
                menuBar.setVisibility(View.GONE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            }
            //Accessing the image gallery
            Intent photoPickerIntent=new Intent(Intent.ACTION_PICK);
            File pictureDirectory= Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
            String pictureDirectoryPath=pictureDirectory.getPath();
            Uri data=Uri.parse(pictureDirectoryPath);
            photoPickerIntent.setDataAndType(data,"image/*");
            startActivityForResult(photoPickerIntent, IMAGE_GALLERY_REQUEST);
        }
        else if(v.equals(menuBtn))
        {
            if(visible==0)
            {
                visible=1;
                menuBar.setVisibility(View.VISIBLE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.cross);
            }
            else
            {
                visible=0;
                menuBar.setVisibility(View.GONE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            }
        }
        else if(v.equals(background))
        {
            if(visible==1) {
                visible = 0;
                menuBar.setVisibility(View.GONE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            }
        }
        else if(v.equals(recyclerMsg))
        {
            if(visible==1) {
                visible = 0;
                menuBar.setVisibility(View.GONE);
                loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            }
        }
        else if(v.equals(item1))
        {
            hideKeyboard(MainActivity.this);
            emojiRel.setVisibility(View.VISIBLE);
            isEmojiPaneVisible=true;
            menuBarParams=(RelativeLayout.LayoutParams)menuBar.getLayoutParams();
            menuBarParams.addRule(RelativeLayout.ABOVE,R.id.emojiRel);
            menuBarParams.setMargins(0,0,0,-1*(int)this.getResources().getDimension(R.dimen.emojiRelMarginTop));
            menuBar.setLayoutParams(menuBarParams);
        }
        else if(v.equals(item2))
        {

        }
        else if(v.equals(item3))
        {

        }
        else if(v.equals(item4))
        {

        }
        else if(v.equals(item5))
        {

        }
    }



    static void setCurrentX(int x)
    {
        currentX=x;
    }
    static void setCurrentY(int y)
    {
        currentY=y;
    }
    static void setStartX(int x)
    {
        startX=x;
    }
    static void setStartY(int y)
    {
        startY=y;
    }

    String getTime()
    {
        Calendar rightNow=Calendar.getInstance();
        int hr=rightNow.get(Calendar.HOUR);
        int hr2=rightNow.get(Calendar.HOUR_OF_DAY);
        int min=rightNow.get(Calendar.MINUTE);
        String half="";
        String time="";
        if(hr2>=12)
        {
            half = "PM";
        }
        else
            half="AM";
        if(min>=10)
            time=Integer.toString(hr)+":"+Integer.toString(min)+" "+half;
        else
            time=Integer.toString(hr)+":0"+Integer.toString(min)+" "+half;
        return(time);
    }

    String getDate()
    {
        Date date=Calendar.getInstance().getTime();
        SimpleDateFormat dateFormat=new SimpleDateFormat("dd-MMM-yyyy");
        String formattedDate=dateFormat.format(date);
        //Toast.makeText(MainActivity.this,""+formattedDate,Toast.LENGTH_SHORT).show();
        return formattedDate;
    }

    @Override
    public void onFocusChange(View v, boolean hasFocus)
    {
        if(v.equals(msg))
        {
            if(hasFocus)
            {
                if(i==0) {
                    firstType.setVisibility(View.VISIBLE);
                    secondType.setVisibility(View.GONE);
                }
                else
                {
                    firstType.setVisibility(View.GONE);
                    secondType.setVisibility(View.VISIBLE);
                }
            }
            else
            {
                firstType.setVisibility(View.GONE);
                secondType.setVisibility(View.GONE);
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == IMAGE_GALLERY_REQUEST) {
                Uri imageUri = data.getData();

                messages.add(imageUri.toString());
                if (i == 0)
                    sender.add(2);
                else
                    sender.add(3);
                sentTime.add(getTime());
                adjustDates();
                msgLayoutManager.scrollToPosition(messages.size() - 1);
                msg.setText("");
                recyclerMsg.requestFocus();

                if (i == 0)
                    i = 1;
                else
                    i = 0;

                msg.requestFocus();
            }
        }
    }

    @Override
    public void onItemClick(int position)
    {
        //Toast.makeText(this,"clicked",Toast.LENGTH_SHORT).show();
        if(visible==1) {
            visible = 0;
            menuBar.setVisibility(View.GONE);
            loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
        }
    }

    public void adjustDates()
    {
        sentDate.add(getDate());
        if(sentDate.size()==1)
        {
            datePosition.add(messages.size()-1);
            showdate.add("Today");
        }
        else if(!sentDate.get(sentDate.size()-1).equals(sentDate.get(sentDate.size()-2)))
        {
            datePosition.add(messages.size()-1);
            showdate.add("Today");
            showdate.set(datePosition.size()-2,"Yesterday");
            if(showdate.size()-3>=0)
                showdate.set(datePosition.size()-3,sentDate.get(datePosition.size()-3));
        }
        else
        {
            showdate.add("");
        }
        msgAdapter.notifyItemInserted(messages.size()-1);
        if(datePosition.size()==1)
        {
            msgAdapter.notifyItemChanged(datePosition.get(0));
        }
        else if(datePosition.size()==2)
        {
            msgAdapter.notifyItemChanged(datePosition.get(0));
            msgAdapter.notifyItemChanged(datePosition.get(1));
        }
        else
        {
            msgAdapter.notifyItemChanged(datePosition.get(datePosition.size()-1));
            msgAdapter.notifyItemChanged(datePosition.get(datePosition.size()-2));
            msgAdapter.notifyItemChanged(datePosition.get(datePosition.size()-3));
        }
    }

    public static void initImages(Context context)
    {
        loadIconOrMenuImg(context,menuBtn,R.drawable.plus);
        loadIconOrMenuImg(context,sendBtn,R.drawable.send_btn);
        loadIconOrMenuImg(context,attachmentBtn,R.drawable.attachment);
        loadIconOrMenuImg(context,item1,R.drawable.right_thumbs_up);
        loadIconOrMenuImg(context,item2,R.drawable.font_logo);
        loadIconOrMenuImg(context,item3,R.drawable.placeholder);
        loadIconOrMenuImg(context,item4,R.drawable.placeholder);
        loadIconOrMenuImg(context,item5,R.drawable.placeholder);


    }

    public static void loadIconOrMenuImg(Context context,ImageButton imgBtn, int image)
    {
        GlideApp
                .with(context)
                .load(image)
                .into(imgBtn);
    }
    public static void hideKeyboard(Activity activity) {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
        //Find the currently focused view, so we can grab the correct window token from it.
        View view = activity.getCurrentFocus();
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = new View(activity);
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    @Override
    public void onEmojiClick(int position)
    {
        if(i==0)
        {
            sender.add(6);
            messages.add(Integer.toString(emojiName.get(position)));
            i=1;
        }
        else
        {
            sender.add(7);
            messages.add(Integer.toString(emojiName.get(position)));
            i=0;
        }
        sentTime.add(getTime());
        adjustDates();
        msgLayoutManager.scrollToPosition(messages.size()-1);
        recyclerMsg.requestFocus();
        emojiRel.setVisibility(View.GONE);
        isEmojiPaneVisible=false;
        menuBar.setVisibility(View.GONE);
        visible=0;
        loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);

        menuBarParams=(RelativeLayout.LayoutParams)menuBar.getLayoutParams();
        menuBarParams.addRule(RelativeLayout.ABOVE,R.id.baseBar);
        menuBarParams.setMargins(0,0,0,0);
        menuBar.setLayoutParams(menuBarParams);
    }

    @Override
    public void onBackPressed()
    {
        if(isEmojiPaneVisible)
        {
            emojiRel.setVisibility(View.GONE);
            isEmojiPaneVisible=false;
            menuBarParams=(RelativeLayout.LayoutParams)menuBar.getLayoutParams();
            menuBarParams.addRule(RelativeLayout.ABOVE,R.id.baseBar);
            menuBarParams.setMargins(0,0,0,0);
            menuBar.setLayoutParams(menuBarParams);
        }
        else if(visible==1)
        {
            menuBar.setVisibility(View.GONE);
            loadIconOrMenuImg(MainActivity.this,menuBtn,R.drawable.plus);
            visible=0;
        }
        else
            super.onBackPressed();
    }
}

File: app/src/main/java/com/example/chatapp/MainActivity.kt
package com.example.chatapp

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.view.MenuItem
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageButton
import android.widget.PopupMenu
import android.widget.RelativeLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.chatapp.EmojiAdapter.OnEmojiListener
import com.wang.avi.AVLoadingIndicatorView
import net.yslibrary.android.keyboardvisibilityevent.KeyboardVisibilityEvent
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(),
        View.OnClickListener,
        OnFocusChangeListener,
        OnEmojiListener,
        PopupMenu.OnMenuItemClickListener {
    private var recyclerMsg: RecyclerView? = null
    private var emojiPane: RecyclerView? = null
    private var msgAdapter: MessageAdapter? = null
    private var emojiAdapter: EmojiAdapter? = null
    private var qtAdapter: MessageAdapter? = null
    private var msgLayoutManager: RecyclerView.LayoutManager? = null
    private var emojiLayoutManager: RecyclerView.LayoutManager? = null
    private var quoteLayoutManager: RecyclerView.LayoutManager? = null
    var firstType: AVLoadingIndicatorView? = null
    var secondType: AVLoadingIndicatorView? = null
    var msg: EditText? = null
    var background: RelativeLayout? = null
    var emojiRel: RelativeLayout? = null
    var quoteBtn: ImageButton? = null
    var d = 0f
    var SCREEN_WIDTH = 0f
    var SCREEN_HEIGHT = 0f
    var isEmojiPaneVisible = false

    var messages: ArrayList<String>? = null
    var sender: ArrayList<Int>? = null
    var sentTime: ArrayList<String>? = null
    var sentDate: ArrayList<String>? = null
    var showdate: ArrayList<String>? = null
    var datePosition: ArrayList<Int>? = null
    lateinit var msgFont: ArrayList<Int>
    var menuBarParams: RelativeLayout.LayoutParams? = null

    lateinit var msgView: RecyclerView
    lateinit var quotes: ArrayList<String>
    lateinit var font: ArrayList<Int>
    lateinit var bg: ArrayList<Int>
    lateinit var fontColor: ArrayList<Int>
    lateinit var rotationAngle: ArrayList<Float>

    val QUOTE_PICKER_ID = 12;
    var i = 0;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        d = this.resources.displayMetrics.density
        SCREEN_WIDTH = windowManager.defaultDisplay.width.toFloat()
        SCREEN_HEIGHT = windowManager.defaultDisplay.height.toFloat()

        messages = ArrayList()
        sender = ArrayList()
        sentTime = ArrayList()
        sentDate = ArrayList()
        showdate = ArrayList()
        datePosition = ArrayList()
        emojiName = ArrayList()
        msgFont = ArrayList<Int>()
        quotes = ArrayList<String>()
        font = ArrayList<Int>()
        bg = ArrayList<Int>()
        fontColor = ArrayList<Int>()
        rotationAngle = ArrayList<Float>()

        msg = findViewById<View>(R.id.msg) as EditText
        sendBtn = findViewById<View>(R.id.sendBtn) as ImageButton
        recyclerMsg = findViewById<View>(R.id.recyclerMsg) as RecyclerView
        emojiPane = findViewById<View>(R.id.emojiPane) as RecyclerView
        firstType = findViewById<View>(R.id.firstType) as AVLoadingIndicatorView
        secondType = findViewById<View>(R.id.secondType) as AVLoadingIndicatorView
        attachmentBtn = findViewById<View>(R.id.attachmentBtn) as ImageButton
        menuBtn = findViewById<View>(R.id.menuBtn) as ImageButton
        background = findViewById<View>(R.id.background) as RelativeLayout
        emojiRel = findViewById<View>(R.id.emojiRel) as RelativeLayout
        setting = findViewById(R.id.setting) as ImageButton
        msgView = findViewById(R.id.recyclerMsg) as RecyclerView
        quoteBtn = findViewById(R.id.quoteBtn) as ImageButton

        quoteLayoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        msgView.layoutManager = quoteLayoutManager
        msgView.setHasFixedSize(true)
        qtAdapter = MessageAdapter(null, sender!!, null, null, null, quotes, font, bg, fontColor, rotationAngle)
        msgView.adapter = qtAdapter

        quoteBtn!!.setOnClickListener(this)

        initImages(this)

        emojiName = EmojiInitializer.initialize()
        firstType!!.visibility = View.GONE
        secondType!!.visibility = View.GONE
        emojiRel!!.visibility = View.GONE

        msgAdapter = MessageAdapter(messages, sender, sentTime, showdate, msgFont,
                quotes, font, bg, fontColor, rotationAngle)
        msgLayoutManager = LinearLayoutManager(this)
        recyclerMsg!!.layoutManager = msgLayoutManager
        recyclerMsg!!.setHasFixedSize(true)
        recyclerMsg!!.adapter = msgAdapter
        (msgLayoutManager as LinearLayoutManager).scrollToPosition(messages!!.size - 1)

        emojiAdapter = emojiName?.let { EmojiAdapter(it) }
        emojiAdapter!!.setEmojiListener(this)
        emojiLayoutManager = GridLayoutManager(this, 5)
        emojiPane!!.layoutManager = emojiLayoutManager
        emojiPane!!.setHasFixedSize(true)
        emojiPane!!.adapter = emojiAdapter
        (emojiLayoutManager as GridLayoutManager).scrollToPosition(0)

        sendBtn!!.setOnClickListener(this)
        msg!!.onFocusChangeListener = this
        attachmentBtn!!.setOnClickListener(this)
        menuBtn!!.setOnClickListener(this)
        background!!.setOnClickListener(this)
        recyclerMsg!!.setOnClickListener(this)
        setting.setOnClickListener(this)


        KeyboardVisibilityEvent.setEventListener(
                this@MainActivity
        ) { isOpen ->
            if (isOpen == false) {
                val current = currentFocus
                current?.clearFocus()
            }
        }
    }

    override fun onClick(v: View) {
        if (v == sendBtn) {
            val message = msg!!.text.toString().trim { it <= ' ' }
            if (message.length > 0) {
                messages!!.add(message)
                sender!!.add(i)
                sentTime!!.add(time)
                msgFont.add(defaultFont)
                adjustDates()

                quotes.add("")
                font.add(-1)
                bg.add(-1)
                fontColor.add(-1)
                rotationAngle.add(-1f)

                msgLayoutManager!!.scrollToPosition(messages!!.size - 1)
                msgAdapter!!.notifyItemInserted(messages!!.size - 1)
                msg!!.setText("")
                recyclerMsg!!.requestFocus()
                i = if (i == 0) 1 else 0
                msg!!.requestFocus()
            }
        } else if (v == attachmentBtn) {
            //Accessing the image gallery
            val photoPickerIntent = Intent(Intent.ACTION_PICK)
            val pictureDirectory = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
            val pictureDirectoryPath = pictureDirectory.path
            val data = Uri.parse(pictureDirectoryPath)
            photoPickerIntent.setDataAndType(data, "image/*")
            startActivityForResult(photoPickerIntent, IMAGE_GALLERY_REQUEST)
        } else if (v == menuBtn) {
            hideKeyboard(this@MainActivity)
            emojiRel!!.visibility = View.VISIBLE
            isEmojiPaneVisible = true
        } else if (v == background) {
        } else if (v == recyclerMsg) {
        } else if (v == setting) {
            PopupMenu(this, v).apply {
                setOnMenuItemClickListener(this@MainActivity)
                inflate(R.menu.options_menu)
                show()
            }
        } else if (v == quoteBtn) {
            val intent = Intent(applicationContext, SelectQuote::class.java)
            startActivityForResult(intent, QUOTE_PICKER_ID)
        }
    }

    val time: String
        get() {
            val rightNow = Calendar.getInstance()
            val hr = rightNow[Calendar.HOUR]
            val hr2 = rightNow[Calendar.HOUR_OF_DAY]
            val min = rightNow[Calendar.MINUTE]
            var half = ""
            var time = ""
            half = if (hr2 >= 12) {
                "PM"
            } else "AM"
            time = if (min >= 10) Integer.toString(hr) + ":" + Integer.toString(min) + " " + half else Integer.toString(hr) + ":0" + Integer.toString(min) + " " + half
            return time
        }

    val date: String
        get() {
            val date = Calendar.getInstance().time
            val dateFormat = SimpleDateFormat("dd-MMM-yyyy")
            //Toast.makeText(MainActivity.this,""+formattedDate,Toast.LENGTH_SHORT).show();
            return dateFormat.format(date)
        }

    override fun onFocusChange(v: View, hasFocus: Boolean) {
        if (v == msg) {
            if (hasFocus) {
                if (i == 0) {
                    firstType!!.visibility = View.VISIBLE
                    secondType!!.visibility = View.GONE
                } else {
                    firstType!!.visibility = View.GONE
                    secondType!!.visibility = View.VISIBLE
                }
            } else {
                firstType!!.visibility = View.GONE
                secondType!!.visibility = View.GONE
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == IMAGE_GALLERY_REQUEST) {
                val imageUri = data!!.data
                messages!!.add(imageUri.toString())
                if (i == 0) sender!!.add(2) else sender!!.add(3)
                msgFont.add(defaultFont)
                sentTime!!.add(time)
                adjustDates()

                quotes.add("")
                font.add(-1)
                bg.add(-1)
                fontColor.add(-1)
                rotationAngle.add(-1f)

                msgLayoutManager!!.scrollToPosition(messages!!.size - 1)
                msgAdapter!!.notifyItemInserted(messages!!.size - 1)
                msg!!.setText("")
                recyclerMsg!!.requestFocus()
                i = if (i == 0) 1 else 0
                msg!!.requestFocus()
            } else if (requestCode == FONT_REQUEST) {
                var intent: Intent = data as Intent
                defaultFont = intent.getIntExtra("selectedFont", 1)
            } else if (requestCode == QUOTE_PICKER_ID) {
                var myIntent: Intent = data as Intent
                var str: String = myIntent.getStringExtra("selectedQuote")!!
                //Toast.makeText(applicationContext,str,Toast.LENGTH_SHORT).show()
                if (i == 0) {
                    sender?.add(8)
                    i = 1
                } else {
                    sender?.add(9)
                    i = 0
                }
                quotes.add(str)
                font.add(myIntent.getIntExtra("font", 1))
                bg.add(myIntent.getIntExtra("selectedBg", 1))
                fontColor.add(myIntent.getIntExtra("selectedFontColor", 1))
                rotationAngle.add(myIntent.getFloatExtra("selectedRotation", 1f))
//                qtAdapter?.notifyItemInserted(quotes.size-1)
//                quoteLayoutManager?.scrollToPosition(quotes.size-1)
                messages!!.add("-1");
                msgFont.add(defaultFont)
                sentTime!!.add(time)
                adjustDates()
                msgAdapter?.notifyItemInserted(messages!!.size - 1)
                msgLayoutManager!!.scrollToPosition(messages!!.size - 1)
                recyclerMsg!!.requestFocus()
            }
        }
    }

    fun adjustDates() {
        sentDate!!.add(date)
        if (sentDate!!.size == 1) {
            datePosition!!.add(messages!!.size - 1)
            showdate!!.add("Today")
        } else if (sentDate!![sentDate!!.size - 1] != sentDate!![sentDate!!.size - 2]) {
            datePosition!!.add(messages!!.size - 1)
            showdate!!.add("Today")
            showdate!![datePosition!!.size - 2] = "Yesterday"
            if (showdate!!.size - 3 >= 0) showdate!![datePosition!!.size - 3] = sentDate!![datePosition!!.size - 3]
        } else {
            showdate!!.add("")
        }
        msgAdapter!!.notifyItemInserted(messages!!.size - 1)
        if (datePosition!!.size == 1) {
            msgAdapter!!.notifyItemChanged(datePosition!![0])
        } else if (datePosition!!.size == 2) {
            msgAdapter!!.notifyItemChanged(datePosition!![0])
            msgAdapter!!.notifyItemChanged(datePosition!![1])
        } else {
            msgAdapter!!.notifyItemChanged(datePosition!![datePosition!!.size - 1])
            msgAdapter!!.notifyItemChanged(datePosition!![datePosition!!.size - 2])
            msgAdapter!!.notifyItemChanged(datePosition!![datePosition!!.size - 3])
        }
    }

    override fun onEmojiClick(position: Int) {
        i = if (i == 0) {
            sender!!.add(6)
            messages!!.add(Integer.toString(emojiName!![position]))
            1
        } else {
            sender!!.add(7)
            messages!!.add(Integer.toString(emojiName!![position]))
            0
        }
        msgFont.add(defaultFont)
        sentTime!!.add(time)
        adjustDates()

        quotes.add("")
        font.add(-1)
        bg.add(-1)
        fontColor.add(-1)
        rotationAngle.add(-1f)

        msgLayoutManager!!.scrollToPosition(messages!!.size - 1)
        recyclerMsg!!.requestFocus()
        emojiRel!!.visibility = View.GONE
        isEmojiPaneVisible = false
        loadIconOrMenuImg(this@MainActivity, menuBtn, R.drawable.right_thumbs_up)
    }

    override fun onBackPressed() {
        if (isEmojiPaneVisible) {
            emojiRel!!.visibility = View.GONE
            isEmojiPaneVisible = false
        } else super.onBackPressed()
    }

    companion object {
        const val IMAGE_GALLERY_REQUEST = 20
        var sendBtn: ImageButton? = null
        var attachmentBtn: ImageButton? = null
        var menuBtn: ImageButton? = null
        lateinit var setting: ImageButton
        var emojiName: ArrayList<Int>? = null
        val FONT_REQUEST = 42
        var defaultFont = R.font.roboto

        fun initImages(context: Context?) {
            loadIconOrMenuImg(context, menuBtn, R.drawable.right_thumbs_up)
            loadIconOrMenuImg(context, sendBtn, R.drawable.send_btn)
            loadIconOrMenuImg(context, attachmentBtn, R.drawable.attachment)
        }

        fun loadIconOrMenuImg(context: Context?, imgBtn: ImageButton?, image: Int) {
            GlideApp
                    .with(context!!)
                    .load(image)
                    .into(imgBtn!!)
        }

        fun hideKeyboard(activity: Activity) {
            val imm = activity.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
            //Find the currently focused view, so we can grab the correct window token from it.
            var view = activity.currentFocus
            //If no view currently has focus, create a new one, just so we can grab a window token from it
            if (view == null) {
                view = View(activity)
            }
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    override fun onMenuItemClick(item: MenuItem?): Boolean {
        return when (item!!.itemId) {
            R.id.changeFont -> {
                val intent: Intent = Intent(applicationContext, SelectFont::class.java)
                startActivityForResult(intent, FONT_REQUEST)
                true
            }
            else -> false
        }
    }
}